sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/ObjectExpressionsSuite.scala

```diff
@@ -556,6 +556,43 @@ class ObjectExpressionsSuite extends SparkFunSuite with ExpressionEvalHelper {
       "java.lang.Integer is not a valid external type for schema of double")
   }
 
+  test("SPARK-49044 ValidateExternalType should return child in error") {
+    val inputObject = BoundReference(0, ObjectType(classOf[Row]), nullable = true)
+    Seq(
+      (true, BooleanType),
+      (2.toByte, ByteType),
+      (5.toShort, ShortType),
+      (23, IntegerType),
+      (61L, LongType),
+      (1.0f, FloatType),
+      (10.0, DoubleType),
+      ("abcd".getBytes, BinaryType),
+      ("abcd", StringType),
+      (BigDecimal.valueOf(10), DecimalType.IntDecimal),
+      (IntervalUtils.stringToInterval(UTF8String.fromString("interval 3 day")),
+        CalendarIntervalType),
+      (java.math.BigDecimal.valueOf(10), DecimalType.BigIntDecimal),
+      (Array(3, 2, 1), ArrayType(IntegerType))
+    ).foreach { case (input, dt) =>
+      val enc = RowEncoder.encoderForDataType(dt, lenient = false)
+      val validateType = ValidateExternalType(
+        GetExternalRowField(inputObject, index = 0, fieldName = "c0"),
+        dt,
+        EncoderUtils.lenientExternalDataTypeFor(enc))
+      checkObjectExprEvaluation(validateType, input, InternalRow.fromSeq(Seq(Row(input))))
+    }
+
+    checkExceptionInExpression[SparkRuntimeException](
+      ValidateExternalType(
```

Comment:
BTW, can you trigger the error from an user code because if you cannot do that, we should consider to convert the error to an internal one.

Reply:
Yes, look like internal because that look like internals:

```
case class ValidateExternalType might be called from:
    | org.apache.spark.sql.catalyst.SerializerBuildHelper::createSerializer  might be called from:
        |  org.apache.spark.sql.catalyst.encoders.ExpressionEncoder::apply
```

in "user like" level we may face this exception in `TextSocketContinuousStream` that actually tutorials only code.

Reply:
This error may be visible from the code : https://github.com/mrk-andreev/example-spark-schema/blob/main/spark_4.0.0/src/test/scala/ErrorMsgSuite.scala

Reply:
Could you use the code to trigger the error, please.

Reply:
Done

It look like integration test that need spark context. So I placed it into `sql/core/src/test/scala/org/apache/spark/sql/catalyst/expressions/ValidateExternalTypeSuite.scala`.

Reply:
Summary. 

* Integration test placed here `sql/core/src/test/scala/org/apache/spark/sql/catalyst/expressions/ValidateExternalTypeSuite.scala` (checkError + spark.createDataFrame)
* Unit test placed here `sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/ObjectExpressionsSuite.scala`

Also because we changed error message some other tests like `sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/encoders/RowEncoderSuite.scala` was also changed